teps/0027-task-execution-status-at-runtime.md

```diff
@@ -0,0 +1,133 @@
+---
+title: task-exec-status-at-runtime
+authors:
+  - "@pritidesai"
+creation-date: 2020-10-15
+last-updated: 2020-10-15
+status: proposed
+---
+
+# TEP-0027: Task Execution Status at Runtime
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+- [Test Plan](#test-plan)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+A `finally` task in a `pipeline` is executed just before exiting the `pipeline`. `Pipeline` executes `finally` tasks
+after all `dag` tasks finish successfully or one of the `dag` task results in failure i.e `finally` tasks are always
+executed in parallel irrespective of success or failure of a `dag` task. Such `finally` tasks are lacking access to the
+execution status of the `dag` tasks at runtime to better design and process success scenarios v/s failure scenarios.
+
+This proposal is enabling a `finally` task to access the execution status of `dag` tasks at runtime to better design
+`pipeline` based on the status.
+ 
+This proposal also enables any `dag` task to access the execution status of other `dag` tasks.
+
+## Motivation
+
+* A `pipeline` author wants to send notification to an external system based on the execution status of the `dag` tasks. 
+* A `pipeline` author wants to rollback if the deployment fails in that environment/cluster/region/provider.
+
+### Goals
+
+The main goal of this proposal is to allow access to the execution status of a task within the `pipeline` at runtime.
+And design this approach in the simplest possible way, without any extra processing. 
+
+An additional but important goal of this proposal is identify the different execution states of a `pipelineTask` and
+how each of those states derived based on `taskRun` and/or `pipelineRun` status.
+
+### Non-Goals
+
+* This proposal is not to expose the execution state of `finally` tasks since its not required.
+ 
+* This proposal is not to expose the entire status of a `taskRun`/`pipelineRun` as metadata at runtime.
+
+## Proposal
+
+Introduce a `param` value `$(context.pipelineRun.Tasks.<pipelineTask>)` which resolves to one of the execution states:
+`Succeeded`, `Failed`, `Skipped`, and `None`
+
+This `param` value is populated and available at the runtime. In the following example, `deployment` is a `dag` task and `rollback` is a `finally` task. The `finally` task, `rollback`, checks the execution status of `deployment` task using context `param` and continues if `deployment` resulted in failure. 
```

Comment:
i think it's worth calling out somewhere that some potential uses of this field can be highly racey, e.g. if you have 2 tasks run in parallel and they try to check each other's status

this wouldn't be the case in finally tasks tho since we know they always execute after the rest of the tasks - which suggests another potential way to constrain this; we could initially support this ONLY in finally tasks. this might also make sense since some values such as Failed will never be present for non-finally tasks

Reply:
The TEP is about supporting this *only* for `finally` so we are good there.

Reply:
It intended to include both `dag` and `finally` but after having the discussion in API WG today, I have further narrowed scope down to just `finally`.